src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -0,0 +1,504 @@
+  }
+
+  KOKKOS_FUNCTION auto computeNextComponent(int component) const
+  {
+    auto const n = out_edges_.extent(0);
+
+    int next_component = labels_(out_edges_(component - n + 1).target);
+    int next_next_component =
+        labels_(out_edges_(next_component - n + 1).target);
+
+    if (next_next_component != component)
+    {
+      // The component's edge is unidirectional
+      return next_component;
+    }
+    // The component's edge is bidirectional, uniquely resolve the bidirectional
+    // edge
+    return KokkosExt::min(component, next_component);
+  }
+
+  KOKKOS_FUNCTION auto computeFinalComponent(int component) const
+  {
+    int prev_component = component;
+    int next_component;
+    while ((next_component = computeNextComponent(prev_component)) !=
+           prev_component)
+      prev_component = next_component;
+
+    return next_component;
+  }
+
+  KOKKOS_FUNCTION void operator()(int i) const
+  {
+    auto const component = labels_(i);
+    auto const final_component = computeFinalComponent(component);
+    labels_(i) = final_component;
+    if (i != component)
+    {
+      return;
+    }
+    auto const n = out_edges_.extent(0);
+    if (i != final_component)
+    {
+      auto const edge = out_edges_(i - n + 1);
+      // append new edge at the "end" of the array (akin to
+      // std::vector::push_back)
+      auto const back =
+          Kokkos::atomic_fetch_inc(&num_edges_()); // atomic post increment
+      edges_(back) = edge;
+    }
+  }
+};
+
+// Merge components and append new edges
+template <class ExecutionSpace, class Labels, class ComponentOutEdges,
+          class Edges, class EdgesCount>
+void updateComponentsAndEdges(ExecutionSpace const &space,
+                              ComponentOutEdges const &component_out_edges,
+                              Labels const &labels, Edges const &edges,
+                              EdgesCount const &num_edges)
+{
+  UpdateComponentsAndEdges<Labels, ComponentOutEdges, Edges, EdgesCount>(
+      space, labels, component_out_edges, edges, num_edges);
+}
+
+// Reverse node leaf permutation order back to original indices
+template <class ExecutionSpace, class BVH, class Edges>
+void finalizeEdges(ExecutionSpace const &space, BVH const &bvh,
+                   Edges const &edges)
+{
+  auto const n = bvh.size();
+  ARBORX_ASSERT(edges.extent(0) == n - 1);
+  Kokkos::parallel_for(
+      "ArborX::MST::finalize_edges",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n - 1),
+      KOKKOS_LAMBDA(int i) {
+        edges(i).source =
+            HappyTreeFriends::getLeafPermutationIndex(bvh, edges(i).source);
+        edges(i).target =
+            HappyTreeFriends::getLeafPermutationIndex(bvh, edges(i).target);
+      });
+}
+
+template <class ExecutionSpace, class BVH, class Labels, class Metric,
+          class Radii>
+void resetSharedRadii(ExecutionSpace const &space, BVH const &bvh,
```

Comment:
As I mentioned above, I'd rather move this inside the `FindComponentNearestNeighbors` class.